src/ray/common/asio/asio_chaos.cc

```diff
@@ -0,0 +1,103 @@
+// Copyright 2017 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ray/common/asio/asio_chaos.h"
+
+#include <cstdlib>
+#include <string_view>
+
+#include "absl/container/flat_hash_map.h"
+#include "absl/strings/numbers.h"
+#include "absl/strings/str_split.h"
+#include "ray/util/logging.h"
+
+namespace ray {
+namespace asio {
+namespace testing {
+
+namespace {
+/*
+  DelayManager is a simple chaos testing framework. Before starting ray, users
+  should set up os environment to use this feature for testing purposes.
+
+  To use this, simply do
+      export RAY_TESTING_ASIO_DELAY_MS="method1=10,method2=20"
+
+   The delay is a random number between 0 and the value. If method equals '*',
+   it will apply to all methods.
+
+   Please check warnings to make sure delay is on.
+*/
+class DelayManager {
+ public:
+  DelayManager() {
+    auto delay_env = std::getenv("RAY_TESTING_ASIO_DELAY_MS");
```

Comment:
Can we use Ray config def here (should never use getenv since it's not thread-safe)?

Reply:
Although it won't make any difference since it's a static obj and initialized the same as RAY config, I'd like to make sure passing envs are consistent